cpp/src/arrow/util/math_internal.cc

```diff
@@ -0,0 +1,130 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+#include "arrow/util/math_internal.h"
+
+#include <cmath>
+
+#include "arrow/util/logging.h"
+
+namespace arrow::internal {
+
+double NormalPPF(double p) {
+  DCHECK(p >= 0.0 && p <= 1.0);
+  if (p == 0.0) {
+    return -HUGE_VAL;
+  }
+  if (p == 1.0) {
+    return HUGE_VAL;
+  }
+
+  // Algorithm from https://doi.org/10.2307/2347330
+  // Wichura, M. J. (1988).
+  // Algorithm AS 241: The Percentage Points of the Normal Distribution.
+  // Journal of the Royal Statistical Society. Series C (Applied Statistics),
+  // 37(3), 477-484.
+  //
+  // Copied from the Rust implementation at https://github.com/ankane/dist-rust/
```

Comment:
If we use code from other product, could you update `LICENSE.txt` too?

See also: https://infra.apache.org/licensing-howto.html#permissive-deps

Reply:
Good point, I added a relevant snippet.